builder/vsphere/common/artifact.go

```diff
@@ -39,9 +44,82 @@ func (a *Artifact) String() string {
 }
 
 func (a *Artifact) State(name string) interface{} {
+	if name == registryimage.ArtifactStateURI {
+		return a.stateHCPPackerRegistryMetadata()
+	}
 	return a.StateData[name]
 }
 
+// stateHCPPackerRegistryMetadata will write the metadata as an hcpRegistryImage
+func (a *Artifact) stateHCPPackerRegistryMetadata() interface{} {
+	labels := make(map[string]interface{})
+	floppyPath, ok := a.StateData["uploaded_floppy_path"].(string)
+	if ok && floppyPath != "" {
+		labels["uploaded_floppy_path"] = floppyPath
+	}
+	labels = a.getVMInfo(labels)
+
+	img, _ := registryimage.FromArtifact(a,
+		registryimage.WithID(a.Name),
+		registryimage.WithRegion(a.Location.String()),
+		registryimage.WithProvider("vsphere"),
+		registryimage.SetLabels(labels),
+	)
+	return img
+}
+
+func (a *Artifact) getVMInfo(labels map[string]interface{}) map[string]interface{} {
+	if dir, err := a.VM.GetDir(); err == nil {
+		labels["vm_dir"] = dir
+	}
+
+	info, err := a.VM.Info("config.annotation", "config.hardware", "runtime.host", "resourcePool", "datastore", "network", "summary")
+	if err != nil || info == nil {
+		log.Printf("[TRACE] error extracting VM metadata: %s", err)
+		return labels
+	}
+	if info.Config != nil {
+		if info.Config.Annotation != "" {
+			// VM description
+			labels["annotation"] = info.Config.Annotation
+		}
+		// Hardware
+		labels["num_cpu"] = fmt.Sprintf("%d", info.Config.Hardware.NumCPU)
+		labels["num_cores_per_socket"] = fmt.Sprintf("%d", info.Config.Hardware.NumCoresPerSocket)
+		labels["memory_mb"] = fmt.Sprintf("%d", info.Config.Hardware.MemoryMB)
+	}
+
+	h := a.VM.NewHost(info.Runtime.Host)
+	hostInfo, err := h.Info("name")
+	if err == nil && hostInfo.Name != "" {
+		labels["host"] = hostInfo.Name
+	}
+
+	p := a.VM.NewResourcePool(info.ResourcePool)
+	poolPath, err := p.Path()
+	if err == nil && poolPath != "" {
+		labels["resurce_pool"] = poolPath
+	}
+
+	dsr := info.Datastore[0].Reference()
+	ds := a.VM.NewDatastore(&dsr)
+	dsInfo, err := ds.Info("name")
+	if err == nil && dsInfo.Name != "" {
+		labels["datastore"] = dsInfo.Name
+	}
+
+	for i, network := range info.Network {
+		key := fmt.Sprintf("network_%d", i)
+		labels[key] = network.String()
+	}
+
+	if a.ContentLibraryConfig != nil {
+		labels["content_library_destination"] = fmt.Sprintf("%s/%s", a.ContentLibraryConfig.Library, a.ContentLibraryConfig.Name)
```

Comment:
@nywilken I'm including the content library destination based on the user's config for that.